### ISSUE TITLE
CMake crashes while configuring (Incorrect use of CMake)

### ISSUE THREAD

**Issue by rumgot** [libcpr/cpr#852]
### Description

When I add a **Cpr** project to my project (according to the instructions in the **README.md**), the following code in the **Cpr** project's **CMakeLists.txt** causes **CMake** to crash during the configuration phase when using the **Ninja Multi-Config** generator on Linux (Ubuntu 20.04 / 22.04):

```cmake
get_property(isMultiConfig GLOBAL PROPERTY GENERATOR_IS_MULTI_CONFIG)
if (NOT isMultiConfig)
    set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS "${ALLOWED_BUILD_TYPES}")
    if (NOT CMAKE_BUILD_TYPE)
        set(CMAKE_BUILD_TYPE Debug CACHE STRING "" FORCE)
    elseif(NOT CMAKE_BUILD_TYPE IN_LIST ALLOWED_BUILD_TYPES)
        message(FATAL_ERROR "Invalid build type: ${CMAKE_BUILD_TYPE}")
    endif()
else()
    unset(CMAKE_BUILD_TYPE)
    foreach(TYPE ${ALLOWED_BUILD_TYPES})
        if (NOT ${TYPE} IN_LIST CMAKE_CONFIGURATION_TYPES)
            # !!! This command results in an error !!!
            list(APPEND CMAKE_CONFIGURATION_TYPES ${TYPE})
        endif()
    endforeach()
endif()
```

According to the discussion here https://gitlab.kitware.com/cmake/cmake/-/issues/24181, the cmake maintainer thinks that setting CMAKE_CONFIGURATION_TYPES in this manner should not happen. That is supposed to be a cache option for users to set.
Therefore, I ask you to remove the modification of the CMAKE_CONFIGURATION_TYPES variable inside the Cpr project.

### Example/How to Reproduce

The error is reproduced in the Ubuntu 20.04 / 22.04 operating system (most likely this error will also occur in other Linux distributions)

To reproduce, you need to create a project with the following **CMakeLists.txt** file:

```cmake
cmake_minimum_required(VERSION 3.20)
project(CMakeBugList)
include(FetchContent)
FetchContent_Declare(cpr GIT_REPOSITORY https://github.com/libcpr/cpr.git
                         GIT_TAG 871ed52d350214a034f6ef8a3b8f51c5ce1bd400) # The commit hash for 1.9.0. Replace with the latest from: https://github.com/libcpr/cpr/releases
FetchContent_MakeAvailable(cpr)
```

Now let's create a build directory and try to configure the project using the **Ninja Multi-Config** generator:

```shell
cmake ../Bug  -G "Ninja Multi-Config"
...
terminate called after throwing an instance of 'std::out_of_range'
  what():  map::at
Aborted (core dumped)
```

### Where did you get it from?

GitHub (branch e.g. master), Other (specify in "Additional Context/Your Environment")

### Additional Context/Your Environment

- OS: Ubuntu 20.04 / 22.04
- Version: master / 1.9.2

**Comment by bradking** [comment]
> the cmake maintainer thinks that setting CMAKE_CONFIGURATION_TYPES in this manner should not happen.

To clarify: certainly CMake should not crash in this case, and that will be addressed over in the CMake issue.

However, cpr's modification of `CMAKE_BUILD_TYPE` and `CMAKE_CONFIGURATION_TYPES` assumes that cpr is the top-most project in the build.  The example code using FetchContent in this issue's description demonstrates one way in which it might not be the top-most project.  The relevant code block could be made conditional using the `PROJECT_IS_TOP_LEVEL` variable from CMake 3.21+, or `CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR` in older CMake versions.

**Comment by rumgot** [comment]
>To clarify: certainly CMake should not crash in this case, and that will be addressed over in the CMake issue.

In the discussion on the link there is a note that CMake, of course, should not crash with an error.

>The relevant code block could be made conditional using the PROJECT_IS_TOP_LEVEL variable from CMake 3.21+, or CMAKE_CURRENT_SOURCE_DIR STREQUAL CMAKE_SOURCE_DIR in older CMake versions.

This needs to be done in the CMakeLists.txt of the Cpr project.

**Comment by rumgot** [comment]
I also want to point out that adding additional configurations is problematic, for example, in the case of building a Cpr project in Qt Creator: when you open CMakeLists.txt, all configurations are not displayed there (see attached image).
Thus, I think that instead of adding additional configurations for building Cpr with different sets of compilation keys, it is better to use a standard set of configurations, and add the necessary compilation keys depending on the passed CMake options.
![screen](https://user-images.githubusercontent.com/22974167/203486959-deeac97b-7b29-464c-aa4c-ce8a28a19244.png)

**Comment by COM8** [comment]
Ping @KingKili

**Comment by COM8** [comment]
Agree, this is an issue. As soon as #853 got merged, I will publish a new 1.9 release including this fix.

**Comment by COM8** [comment]
https://github.com/libcpr/cpr/releases/tag/1.9.3
